<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="DiversiTools : Tool for analysing diversity from HTS">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>DiversiTools</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/josephhughes/DiversiTools">View on GitHub</a>

          <h1 id="project_title">DiversiTools</h1>
          <h2 id="project_tagline">Tool for analysing viral diversity from HTS</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/josephhughes/DiversiTools/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/josephhughes/DiversiTools/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>This programs has been developed as a result of working with virologist and microbiologist who are interested in understand intra-host diversity of viruses and bacteria and their evolution on a micro-evolutionary scale.
A lot of tools exist that determine the frequency of of mutations (SNPs, indels) from SAM files but these have primarily been developed for diploid model organisms. DiversiTools on the other hand focuses on determining mutations from a population of haploid organisms. 
The large amount of sequence data provided by HTS (High-throughput sequencing) enables us to study virus and bacterial populations beyond the consensus.</p>

<p>To learn to use DiversiTools have a run through the different tutorials:</p>
<p><ul>
<li><a href="tutorial.html">Tutorial for command-line use</a></li>
<li><a href="GUItutorial.html">Tutorial for the Graphical User interface</a></li>
</ul>
</p>

<h3>Papers using DiversiTools (previously btctools)</h3>
<p><ul>
<li>Claire L. Donald , Benjamin Brennan , Stephanie L. Cumberworth , Veronica V. Rezelj, Jordan J. Clark, Marli T. Cordeiro, 
Rafael Freitas de Oliveira França, Lindomar J. Pena, Gavin S. Wilkie, Ana Da Silva Filipe, Christopher Davis, Joseph Hughes, 
Margus Varjak, Martin Selinger, Luíza Zuvanov, Ania M. Owsianka, Arvind H. Patel, John McLauchlan, Brett D. Lindenbach, 
Gamou Fall, Amadou A. Sall, Roman Biek, Jan Rehwinkel, Esther Schnettler, Alain Kohl. 
<b>Full Genome Sequence and sfRNA Interferon Antagonist Activity of Zika Virus from Recife, Brazil.</b> PLoS Negl Trop Dis. 2016 Oct; 10(10): e0005048.
 doi: <a href="http://dx.doi.org/10.1371/journal.pntd.0005048">10.1371/journal.pntd.0005048</a>. 
</li>
<li>Michael Jacobs, Alison Rodger, David J Bell, Sanjay Bhagani, Ian Cropley, Ana Filipe, Robert J Gifford, Susan Hopkins, 
Joseph Hughes, Farrah Jabeen, Ingolfur Johannessen, Drosos Karageorgopoulos, Angie Lackenby, Rebecca Lester, 
Rebecca S N Liu, Alisdair MacConnachie, Tabitha Mahungu, Daniel Martin, Neal Marshall, Stephen Mepham, Richard Orton, 
Massimo Palmarini, Monika Patel, Colin Perry, S Erica Peters, Duncan Porter, David Ritchie, Neil D Ritchie, 
R Andrew Seaton, Vattipally B Sreenu, Kate Templeton, Simon Warren, Gavin S Wilkie, Maria Zambon, Robin Gopal, Emma C Thomson. 
<b>Late Ebola virus relapse causing meningoencephalitis: a case report. </b>
Lancet. 2016 Jul 30; 388(10043): 498–503. doi: <a href="http://dx.doi.org/10.1016/S0140-6736(16)30386-5">10.1016/S0140-6736(16)30386-5</a>.
</li>
<li>Monne I, Fusaro A, Nelson MI, Bonfanti L, Mulatti P, Hughes J, Murcia PR,
Schivo A, Valastro V, Moreno A, Holmes EC, Cattoli G. <b>Emergence of a highly
pathogenic avian influenza virus from a low-pathogenic progenitor.</b> J Virol. 2014 
Apr;88(8):4375-88. doi: <a href="https://doi.org/10.1128/JVI.03181-13">10.1128/JVI.03181-13</a>. Epub 2014 Feb 5. PubMed PMID:
24501401.</li>
<li>Fusaro A, Tassoni L, Hughes J, Milani A, Salviato A, Schivo A, Murcia PR,
Bonfanti L, Cattoli G, Monne I. <b>Evolutionary trajectories of two distinct avian
influenza epidemics: Parallelisms and divergences.</b> Infect Genet Evol. 2015
Aug;34:457-66. doi: <a href="https://doi.org/10.1016/j.meegid.2015.05.020">10.1016/j.meegid.2015.05.020</a>. Epub 2015 May 20. PubMed PMID: 
26003682.</li>
</ul>
</p>




<h3>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>This project is a collaboration between <a href="https://github.com/rjorton" class="user-mention">@rjorton</a> and <a href="https://github.com/josephhughes" class="user-mention">@josephhughes</a> to enable virologists and microbiologists to 
understand the intra-host variation from their high-throughput reads.</p>

      </section>
    </div>


    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">DiversiTools maintained by <a href="https://github.com/josephhughes">josephhughes</a> and <a href="https://github.com/rjorton">rjorton</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
