<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Diversitools : Tool for analysing diversity from Sequence Alignment/Map">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Diversitools</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/josephhughes/DiversiTools">View on GitHub</a>

          <h1 id="project_title"><a href="index.html" style="text-decoration: none; color: white">Diversitools</a></h1>
          <h2 id="project_tagline">Tool for analysing diversity from Sequence Alignment/Map</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/josephhughes/DiversiTools/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/josephhughes/DiversiTools/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>Here I will try to explain how you can run the various diversitools from the command-line. The core of diversitools is written in perl and makes use of the the <a href="http://search.cpan.org/~lds/Bio-SamTools/">Bio-SamTools module</a> written by Lincoln D. Stein. 
        So the various perl script can be run from the command-line to automate the processing on multiple bam files.</p>
        
        <p>There are two basic processing scripts: <a href="#diversiutils">diversiutils</a> which is used for processing a BAM file and <a href="#diversifilter">diversifilter</a> which is used for filtering the output according to quality and strand bias criteria.
     
<h2>
<a name="diversiutils" class="anchor" href="#diversiutils"><span class="octicon octicon-link"></span></a>diversiutils</h2>

<p>This script is used to generate the files that contain the frequencies of various mutations from the bam file. In the most basic form, you may just be interested in the number of SNPs at each site of your alignment. From the directory where the perl script is located:</p>

<pre><code>$ perl diversiutils.pl -bam path/to/input.bam -ref path/to/reference.fasta -stub out
</code></pre>

This is an example using files in the test directory:
<pre><code>$ perl diversiutils.pl -bam test/S18_refLPAI.bam -ref test/refLPAI.fa -stub S18
</code></pre>


<p>Bear in mind that the bam file needs its associated index file (.bai). All of the arguments above are necessary to run diversiutils. The output from the latter command will be 3 files:</p>
<p><ul>
<li>out_entropy.txt which contains the frequency of mutations at each individual site for each of the gene segments</li>
<li>out_log.txt which contains the distribution of mutations per read and the average entropy per gene segment</li>
<li>out_read.txt the per read position mutation count and average quality</li>
</ul>
</p>

<h3>out_entropy.txt file</h3>
<p>out_entropy.txt contains the following columns:
<ul>
<li>Sample: the name of the bam file that has been run</li>
<li>Chr: the gene or chromosome identifier from the reference used for the mapping</li>
<li>Position: the site/position of alignment in the reference</li>
<li>RefBase: the base of the reference sequence at that position</li>
<li>Coverage: the coverage, i.e. number of reads mapping at that position</li>
<li>AvQual: the average base-calling error probability which is related to the Phred quality score</li>
<li>Acnt: the total number of A nucleotides at that position</li>
<li>Apval: average probability of sequencing error for that nucleotide based on quality scores</li>
<li>The values for Ccnt, Cpval, Tcnt, Tpval, Gcnt, Gpval are also listed</li>
<li>entropy(base e): a measure of uncertainty in the dataset used as a means to quantify sequence variability at that particular site</li>
<li>NonRefCnt: the number of reads that have bases different to the reference</li>
<li>CntTv: the total number of transversions at that site</li>
<li>CntTs: the total number of transitions at that sites</li>
<li>OrderOfNucs: the order of the nucleotides based on each nucleotide count</li>
<li>Strandbias: the number of reads supporting each base in the forward and the reverse direction (e.g., A1:0;C83:2;T0:0;G1:0)</li>
<li>ins_cnt: the number of reads with inserts at that position</li>
<li>ins_mode: the sequence of the insert that occurs the most frequently</li>
<li>del_cnt: the number of reads with deletions at that site</li>
<li>del_mode: the sequence of the deletion that occurs the most frequently</li>
</ul>
</p>

<h3>out_log.txt file</h3>
<p>out_log.txt contains summary statistics from the alignment:
<ul>
<li>Frequency of mismatches per read (mismatches: number of reads): this provides a distribution of the number of mismatches found per read</li>
<li>Gene Average entropy: the average entropy for each of the genes in the dataset is provided</li>
</ul>
</p>

<h3>out_read.txt file</h3>
<p>out_read.txt contains information about mismatches per read position which can be used to determine whether harsher trimming is required. This table assumes that you have 
the same number of bases in each read, if not the results may be irrelevant as there will be different coverage for each site of the reads:
<ul>
<li>ReadPos: the position in the read (e.g., 1 to 150)</li>
<li>CntRef: number of bases matching the reference at this position</li>
<li>CntNonRef: number of bases not matching the reference at that position in the read</li>
<li>TotalCnt: total number of bases (ref + nonref)</li>
<li>Freq: frequency of mismatches at this read position</li>
<li>AvQual: the average base-calling error probability for that position in a read</li>
<li>AvQualRef: the average base-calling probability for nucleotides matching the reference</li>
<li>AvQualNonRef: the average base-calling probability for frequency for nucleotides different to the reference</li>
</ul>
</p>

<h2>
<a name="diversiutils" class="anchor" href="#diversiutils"><span class="octicon octicon-link"></span></a>diversiutils with ORFs</h2>


<p>It is also possible to provide information about the number of non-synonymous and synonymous mutations at each amino acid site, but to do this you need to provide a text file (e.g. CodingRegion.txt) with the information about the protein names and open reading frames, in the following tab delimited format:</p>
<pre><code>
Protein	Beg	End	Reference
PB2	28	2307	PB2_Influenza
PB1	25	2301	PB1_Influenza
PB1-F2	119	391	PB1_Influenza
PA	25	2176	PA_Influenza
PA-X_A	25	596	PA_Influenza
PA-X_B	598	784	PA_Influenza
</code></pre>

<p>The first column corresponds to the protein name, the names in this column must be unique, the second column corresponds to the start of the orf for that protein, the third is the end of the orf for the protein and the fourth corresponds to the identifier (name) of the reference in your reference file from which the protein is transcribed. Run the command:</p>

<pre><code>$ perl btcutils.pl -bam path/to/input.bam -ref path/to/reference.fasta -orfs CodingRegions.txt -stub out
</code></pre>

<p>The latter command produces all of the previous 4 files with an additional out_AA.txt with the information about the non-synonymous, synonymous and stop-codons.</p>

<h3>out_AA.txt file</h3>
<p>out_AA.txt contains information about the codons and amino acids at each site of the coding regions for each protein. The file contains the following columns:
<ul>
<li>Sample: name of the sample bam file</li>
<li>Chr: name of the chromosome or gene based on the reference used for the alignment</li>
<li>Protein: name of the protein based on the first column of the coding region information file (e.g., CodingRegion.txt)</li>
<li>AAPosition: the amino acid position from start position of the protein</li>
<li>RefAA: the reference amino acid</li>
<li>RefSite: the reference position for the first codon position</li>
<li>RefCodon: the reference codon</li>
<li>FstCodonPos: the number of mutations in the first codon position</li>
<li>SndCodonPos: the number of mutations in the second codon position</li>
<li>TrdCodonPos: the number of mutations in the third codon position</li>
<li>CntNonSyn: the number of non synonymous changes at the amino acid site</li>
<li>CntSyn: the number of synonymous changes at the amino acid site</li>
<li>NbStop: the number of stop codons at this amino acid site</li>
<li>TopAA: the top (most frequently  found) amino acid (usually the reference amino acid)</li>
<li>TopAAcnt: the number of times this amino acid was found</li>
<li>SndAA: the second most frequently found amino acid</li>
<li>SndAAcnt: the number of times this amino acid was found</li>
<li>TrdAA: the third most frequent amino acid (this could go on for ever ...)</li>
<li>TrdAAcnt: the number of times the third most frequent amino acid was found</li>
<li>AAcoverage: the amino acid coverage which may be a bit lower than the nucleotide coverage as it only takes into account full codons</li>
</ul>
</p>

<h2>
<a name="multisamples" class="anchor" href="#multisampels"><span class="octicon octicon-link"></span></a>Processing multiple samples</h2>

<p>One of the advantages of being able to run scripts from the command line, is the option to be able to automate the process. 
For example, we can process all the bam files that have been aligned to the LPAI reference wil the following script.</p>

<pre><code>$ for file in test/*_refLPAI.bam;
do 
perl diversiutils.pl -bam $file -ref test/refLPAI.fa -orf test/Coding.regions.LPAI.txt -stub ${file%_refLPAI.bam};
done
</code></pre>


<h2>
<a name="diversifilter" class="anchor" href="#diversifilter"><span class="octicon octicon-link"></span></a>diversifilter</h2>

<p>Once you have produced processed multiple files, you may want to filter them based on the probabilities of the bases being above the significance level of the binomial test or to remove bases that have strand bias using the Fisher's exact test. For this you can use the diversifilter perl script.</p>

<p>This is an example of filtering the entropy file according to bases that are found below the quality threshold of 0.05 and below the strand bias significance of 0.05.</p>
<pre><code>$ perl diversifilter.pl -in S18 -pQ 0.05 -pS 0.05 -stub S18_pQ05pS05
</code></pre>
<p>The output is the same as the out_entropy.txt file except that base counts that have not passed the threshold will be replaced with 0 and the entropy will be recalculated using only the bases that have passed the quality threshold.</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">diversitools maintained by <a href="https://github.com/josephhughes">josephhughes</a> and <a href="https://github.com/rjorton">rjorton</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
